sources/api/shibaken/src/main.rs

```diff
@@ -0,0 +1,318 @@
+/// embed it into the endpoint url.
+fn get_public_key_endpoints_from_imds(public_key_list: &str) -> Result<Vec<Url>> {
+    let mut public_key_endpoints = Vec::new();
+    for available_key in public_key_list.lines() {
+        let f: Vec<&str> = available_key.split('=').collect();
+        // If f[0] isn't a number, then it isn't a valid index.
+        if f[0].parse::<i32>().is_ok() {
+            let public_key_endpoint = &format!("{}{}/openssh-key", IMDS_PUBLIC_KEY_BASE_URL, f[0]);
+            let public_key_endpoint_url =
+                Url::parse(&public_key_endpoint).context(error::HttpUrlParse {
+                    url: public_key_endpoint,
+                })?;
+            debug!("{}", &public_key_endpoint);
+            public_key_endpoints.push(public_key_endpoint_url);
+        } else {
+            warn!(
+                "'{}' does not appear to be a valid index. Skipping...",
+                &f[0]
+            );
+            continue;
+        }
+    }
+    ensure!(
+        !public_key_endpoints.is_empty(),
+        error::NoValidImdsKeyEndpoints
+    );
+    Ok(public_key_endpoints)
+}
+
+/// Print usage message.
+fn usage() -> ! {
+    let program_name = env::args().next().unwrap_or_else(|| "program".to_string());
+    eprintln!("Usage: {} user-data", program_name);
+    process::exit(2);
+}
+
+/// Parses args for the setting key name.
+fn parse_args(mut args: env::Args) -> String {
+    let arg = args.nth(1).unwrap_or_else(|| usage());
+    if arg == "--help" || arg == "-h" {
+        usage()
+    }
+    arg
+}
+
+async fn run() -> Result<()> {
+    let setting_name = parse_args(env::args());
+
+    // TerminalMode::Mixed will send errors to stderr and anything less to stdout.
+    TermLogger::init(
+        LevelFilter::Debug,
+        LogConfig::default(),
+        TerminalMode::Mixed,
+    )
+    .context(error::Logger)?;
+
+    info!("shibaken started");
+
+    info!("Fetching user-data metadata");
+    let metadata_user_data = get_metadata(&setting_name, "user-data").await?;
+
+    let output = if metadata_user_data.is_empty() {
+
+        info!("No public keys available in user-data. Checking IMDS...");
+        // If there are no public_keys in user-data, then we'll need to fetch authorized_keys from IMDS.
+        let public_keys = get_public_keys_from_imds()?;
+
+        // Generate user_data struct.
+        let ssh_settings = Ssh {
+            authorized_keys: public_keys,
+        };
+
+        let user_data = UserData { ssh: ssh_settings };
+
+        info!("Generating user-data");
+        // Serialize user_data to a JSON string.
+        let user_data_json = serde_json::to_string(&user_data).context(error::SerializeJson)?;
+        debug!("{}", &user_data_json);
+
+        info!("Encoding output");
+        // Encode the output as base64.
+        base64::encode(&user_data_json);
+
+    } else {
+        metadata_user_data;
+    };
```

Comment:
I don't think we want this `else` condition here.  If the user-data exists, we don't need to do anything.